nixos/modules/services/security/opensnitch.nix

```diff
@@ -0,0 +1,174 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.services.opensnitch;
+  inherit (pkgs) opensnitchd opensnitch-ui writeText iptables;
+  inherit (lib) nameValuePair mapAttrsToList;
+  inherit (builtins) listToAttrs replaceStrings concatMap;
+  escapeRuleName = name:
+    replaceStrings [ "*" "." "/" ] [ "all" "-" "-" ] name;
+
+  makeAlwaysRuleFile =  rulePath: name': rule: let name = escapeRuleName name'; in
+    nameValuePair (rulePath + "/${name}.json") {
+      source = writeText name (builtins.toJSON ({
+        inherit name;
+        enabled = true;
+        duration = "always";
+      } // rule));
+    };
+
+  makeAlwaysOperatorFile = rulePath: name: operator:
+    makeAlwaysRuleFile rulePath name {action = "allow"; inherit operator; };
+
+  makePackageRuleFile = rulePath: pkg: let
+    name = "nixos-allow-pkg-${lib.strings.getName pkg}";
+    in makeAlwaysOperatorFile rulePath name {
+          type = "regexp";
+          operand = "process.path";
+          data = "${pkg}/.*";
+    };
+
+  makeHostRuleFile = rulePath: type: value: let
+    name = "nixos-allow-${type}-${value}";
+    in makeAlwaysOperatorFile rulePath name {
+      type = "regexp";
+      operand = "dest.${type}";
+      data = replaceStrings [ "." "*" ] [ "\\." ".*" ] value;
+    };
+
+in
+
+{
+
+  options = {
+    services.opensnitch = with lib; {
+      enable = mkEnableOption "opensnitch";
+
+      uiConfig = mkOption {
+        type = types.attrs;
+        description = "JSON attribute set for opensnitch-ui config file.";
+        default = {
```

Comment:
Better to set this under `config`. If the sysadmin specifies any value under `default` them all default values will be lost.

Reply:
What is the best way to do this? Using ``defaultText`` instead of ``default`` and somehow putting the default values into the ``config``-section?

Reply:
There are 2 main ways to do it:
- use the `apply` attribute in `mkOption`
- don't set a `default` in `mkOption`, instead using `config.services.opensnitch.settings`, and use sufficient [hand-waving](https://en.wikipedia.org/wiki/Hand-waving)

Generally speaking, the entire NixOS module system is pretty opaque unless you look at the code anyways, so the second option isn't as bad as it sounds.